src/System.Data.SqlClient/src/System/Data/SqlClient/TdsParserStateObject.cs

```diff
@@ -3049,16 +3053,36 @@ internal Task WriteByteArray(byte[] b, int len, int offsetBuffer, bool canAccumu
                                 completion = new TaskCompletionSource<object>();
                                 task = completion.Task; // we only care about return from topmost call, so do not access Task property in other cases
                             }
-                            WriteByteArraySetupContinuation(b, len, completion, offset, packetTask);
+
+                            if (array == null)
+                            {
+                                byte[] tempArray = new byte[len];
```

Comment:
Can this tempArray be stackalloc'ed instead of using `new`

Reply:
No, because if the code has reached this point then the source data is in a span and there is no array source, so we've only got a span. At this point we have to pass the data into the async continuation so we can't use a span we have to have the data heap allocated to carry on so copy only the remaining data into a new array and pass that through as the array parameter. When the continuation calls back into the method again it will have an array parameter and will use it from that point onwards.

It changes the flow from always requiring an array to letting you _try_ to use a span but handling the write resume into the next packet if your span wouldn't fit.

Reply:
Got it. Thanks. I think there is a pattrn for when the array is used vs when Span is used for this method. I wish the code structure can make the usage of Span and byte[] clear. I think the method warrants at least a comment explaining when to use what, in case refactoring is not possible to clarify it.

Reply:
Rent with `ArrayPool<byte>.Shared>Rent` then return later if you are using an array param (which is rented)?

Reply:
In order to know if the array is a rental or not I've got to pass that information around. I'm also wary of overusing the array pool because I can see it becoming point of contention if everyone chooses to use the pool instead of new. In this case the spans passed should be no more than 16 bytes long in the cases I can see, do you think it's worth using the pool for that?

Reply:
Probably not; just mentioning it :)

Reply:
It is something I considered and if this area turns up on profiles I'll certainly give it a try.